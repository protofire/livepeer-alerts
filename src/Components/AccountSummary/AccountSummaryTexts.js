export const LOADING_USER_DATA = 'Loading user data'
export const WELCOME_AGAIN = 'Welcome Again! '
export const WELCOME_NEW_SUBSCRIBER = 'Welcome to Livepeer! '
export const WELCOME_NOT_SUBSCRIBED = 'Welcome to Livepeer!'
export const FAIL_NO_REASON = 'Something went wrong!, please try again later!'
export const FAIL_NO_REASON_REDIRECT = 'Something went wrong! Redirecting to home'
export const BOUNDED_STATUS_NEEDED = 'In order to subscribe you need to be on BOUNDED status'
export const LOADING_UNSUBSCRIPTION = 'Please wait while we process your unsubscription'
export const LOADING_SUBSCRIPTION = 'Please wait while we process your subscription'
export const LOADING_SUBSCRIPTION_DATA =
  'Please wait while we get the information of your subscription'
export const GENERATING_SUBSCRIPTION = 'Please wait while we generate your subscription'
export const UNSUBSCRIPTION_SUCCESSFUL = 'You were successfully unsubscribed!'
export const EMAIL_ALREADY_EXISTS = 'Email related to that address already subscribed'
export const SUBSCRIPTION_SUCCESSFUL = 'You were successfully subscribed'
export const FAIL_EMAIL_ALREADY_EXISTS_RESPONSE = 'Subscriptor already exist'
