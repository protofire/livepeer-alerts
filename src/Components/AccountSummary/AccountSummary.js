import * as displayTexts from './AccountSummaryTexts'
import AccountSummaryHome from './AccountSummaryHome/AccountSummaryHome'
import React, { Component } from 'react'
import SpinnerExtended from '../Common/UI/SpinnerExtended/SpinnerExtended'
import { toast, ToastContainer } from 'react-toastify'
import logdown from 'logdown'
import ReactGA from 'react-ga'
import axiosInstance from '../../util/axios'
import { getTranscoderRoi } from '../../util/delegates-utils'

const logger = logdown('Livepeer:AccountSummary')
logger.state.isEnabled = process.env.NODE_ENV !== 'production'

export class AccountSummaryComponent extends Component {
  state = {
    userData: {
      address: null,
      isSubscribed: false,
      activated: null,
      id: null,
      email: '',
      frequency: 'daily',
      activatedCode: null,
      createdAt: null
    },
    summary: {
      role: '',
      lpBalance: '',
      roi: '',
      delegator: {
        address: '',
        allowance: '0',
        bondedAmount: '0',
        delegateAddress: '',
        delegatedAmount: '0',
        fees: '0',
        lastClaimRound: '0',
        pendingFees: '0',
        pendingStake: '0',
        startRound: '0',
        status: '',
        withdrawRound: '0',
        withdrawAmount: '0',
        nextUnbondingLockId: '0',
        totalStake: '0',
        delegateCalledReward: false,
        totalStakeInLPT: '0',
        bondedAmountInLPT: '0'
      },
      delegate: {
        active: true,
        address: '',
        feeShare: '',
        lastRewardRound: '',
        pricePerSegment: '',
        pendingRewardCut: '',
        pendingFeeShare: '',
        pendingPricePerSegment: '',
        rewardCut: '',
        status: '',
        totalStake: '',
        delegateCalledReward: false,
        totalStakeInLPT: ''
      }
    },
    render: false,
    displayMsg: displayTexts.LOADING_USER_DATA,
    toastId: 1,
    error: false
  }

  componentWillReceiveProps(nextProps, nextContext) {
    logger.log('Fire event componentWillReceiveProps', nextProps)
    let propsChanged =
      this.props.render !== nextProps.render ||
      this.props.userData.authenticated !== nextProps.userData.authenticated ||
      this.props.userData.address !== nextProps.userData.address ||
      this.props.userData.currentNetwork !== nextProps.userData.currentNetwork
    if (propsChanged) {
      this.setState(
        {
          ...this.state,
          ...nextProps,
          render: false
        },
        async () => {
          await this.loadUserData()
        }
      )
    }
  }

  loadUserData = async () => {
    let userDataPromise, summaryPromise, roiPromise
    // Check if the user is subscribed
    userDataPromise = this.fetchSubscriptionData()
    // Get summary information about the user
    summaryPromise = this.fetchAccountSummaryData()
    // Get ROI data about the delegate
    roiPromise = this.fetchRoiData()
    try {
      await Promise.all([userDataPromise, summaryPromise, roiPromise])
      this.setState(
        {
          ...this.state,
          render: true,
          displayMsg: displayTexts.WELCOME_AGAIN + this.state.userData.email
        },
        () => logger.log('Loading userData finished', this.state)
      )
    } catch (exception) {
      logger.log('exception ', exception)
    }
  }

  initState = callback => {
    let address = this.props.userData ? this.props.userData.address : ''
    const stringAddresses = process.env.REACT_APP_DEMO_ADDRESS
    // If we are on demo version we choose a random address
    if (this.props.location && this.props.location.pathname === '/account/demo') {
      if (typeof stringAddresses !== 'undefined' && stringAddresses.length > 0) {
        // We have addresses on the env var **/
        const addresses = stringAddresses && stringAddresses.split(',')
        address = addresses && addresses[Math.floor(Math.random() * addresses.length)]
      } else {
        // The user does not have addresses on the env var, we redirect to the home page
        this.props.history.push('/')
      }
    }
    this.setState(
      {
        userData: {
          ...this.state.userData,
          address: address,
          ethBalance: this.props.userData.ethBalance,
          authenticated: this.props.userData.authenticated
        }
      },
      callback
    )
  }

  componentDidMount = async () => {
    logger.log('Fire event componentDidMount')
    // Google analytics
    if (this.props.location && this.props.location.pathname) {
      logger.log('Google analytics: ', this.props.location.pathname)
      ReactGA.pageview(this.props.location.pathname)
    }
    let userDataPromise, summaryPromise
    this.initState(async () => {
      // Check if the user is subscribed
      userDataPromise = this.fetchSubscriptionData()
      // Get summary information about the user
      summaryPromise = this.fetchAccountSummaryData()
      try {
        await Promise.all([userDataPromise, summaryPromise])
        this.setState(
          {
            ...this.state,
            render: true,
            displayMsg: displayTexts.WELCOME_AGAIN + this.state.userData.email
          },
          async () => {
            // Get ROI data about the delegate
            await this.fetchRoiData()
            logger.log('ComponentDidMountFinished ')
          }
        )
      } catch (exception) {
        logger.log('Exception ', exception)
        this.setState(
          {
            ...this.state,
            render: true,
            error: true,
            displayMsg: displayTexts.FAIL_NO_REASON_REDIRECT
          },
          () => {
            this.sendToast(1500, () => this.props.history.push('/'))
          }
        )
      }
    })
  }

  fetchSubscriptionData = async () => {
    return new Promise(async (resolve, reject) => {
      let userData
      try {
        logger.log('Retrieving subscription for address ', this.state.userData.address)
        userData = await axiosInstance.get('/address/' + this.state.userData.address)

        this.setState(
          {
            userData: {
              ...this.state.userData,
              isSubscribed: true,
              id: userData.data._id,
              ...userData.data
            }
          },
          () => {
            resolve(this.state)
          }
        )
      } catch (error) {
        // Subscription not found
        if (error.response && error.response.status === 404) {
          logger.log('Subscription not found for ', this.state.userData.address)
          this.setState(
            {
              ...this.state,
              isSubscribed: false
            },
            () => resolve(this.state)
          )
        } else {
          // Another network problem
          reject(error)
        }
      }
    })
  }

  fetchAccountSummaryData = async () => {
    return new Promise(async (resolve, reject) => {
      try {
        logger.log('Retrieving account summary data for address ', this.state.userData.address)
        let summaryData = await axiosInstance.get('/summary/' + this.state.userData.address)
        this.setState(
          {
            summary: {
              role: summaryData.data.role,
              lpBalance: summaryData.data.balance,
              delegate: summaryData.data.transcoder ? { ...summaryData.data.transcoder } : null,
              delegator: summaryData.data.delegator ? { ...summaryData.data.delegator } : null
            }
          },
          () => resolve(summaryData.data)
        )
      } catch (exception) {
        reject(exception)
      }
    })
  }

  fetchRoiData = async () => {
    const { summary } = this.state
    logger.log('Retrieving ROI data for delegate address ')
    const delegateAdd = summary.delegate
      ? summary.delegate.address
      : summary.delegator.delegateAddress
    if (delegateAdd) {
      const roi = await getTranscoderRoi(delegateAdd)
      this.setState({
        ...this.state,
        summary: {
          ...this.state.summary,
          roi
        }
      })
    }
  }

  sendToast = (toastTime, callback) => {
    let time = 6000
    if (toastTime) {
      time = toastTime
    }
    let displayMsg = this.state.displayMsg

    if (!toast.isActive(this.state.toastId) && this.state.render) {
      if (this.state.error) {
        toast.error(displayMsg, {
          position: toast.POSITION.TOP_RIGHT,
          progressClassName: 'Toast-progress-bar',
          autoClose: time,
          toastId: this.state.toastId,
          onClose: callback
        })
      } else {
        toast.success(displayMsg, {
          position: toast.POSITION.TOP_RIGHT,
          progressClassName: 'Toast-progress-bar',
          autoClose: time,
          toastId: this.state.toastId,
          onClose: callback
        })
      }
    }
  }

  onSubscribeBtnHandler = async () => {
    this.props.history.push('/account/subscription')
  }

  onUnSubscribeBtnHandler = async () => {
    logger.log('Unsubscribe btnHandler')
    this.setState({
      render: false,
      displayMsg: displayTexts.LOADING_UNSUBSCRIPTION
    })
    try {
      logger.log('Unsubscribing user with id ', this.state.userData)
      await axiosInstance.delete('/' + this.state.userData.id)
      this.setState(
        {
          render: true,
          displayMsg: displayTexts.UNSUBSCRIPTION_SUCCESSFUL,
          userData: {
            ...this.state.userData,
            isSubscribed: false,
            activated: null,
            id: null,
            activatedCode: null,
            createdAt: null,
            error: false
          }
        },
        () => this.sendToast()
      )
    } catch (exception) {
      logger.log('Exception on deleteSubscription')
      if (exception.response.status === 404) {
        // User with that id not found
        this.setState(
          {
            render: true,
            displayMsg: displayTexts.WELCOME_NOT_SUBSCRIBED,
            error: true
          },
          () => this.sendToast()
        )
      } else {
        this.setState(
          {
            render: true,
            displayMsg: displayTexts.FAIL_NO_REASON,
            error: true
          },
          () => this.sendToast()
        )
      }
    }
  }

  render() {
    let content = <SpinnerExtended displayMsg={this.state.displayMsg} />
    const { summary } = this.state
    // Shows only summary information according the role (delegate or delegator)
    let summaryForRole = summary.delegate ? summary.delegate : summary.delegator
    let summaryProps = {
      ...summaryForRole,
      roi: summary.roi,
      role: summary.role,
      balance: summary.lpBalance
    }

    if (this.state.render) {
      if (!this.state.error) {
        content = (
          <>
            <AccountSummaryHome
              onUnSubscribeBtnHandler={this.onUnSubscribeBtnHandler}
              onSubscribeBtnHandler={this.onSubscribeBtnHandler}
              web3={this.props.web3}
              userData={this.state.userData}
              summary={summaryProps}
              lpBalance={this.state.summary.lpBalance}
            />
          </>
        )
      } else {
        content = <SpinnerExtended displayMsg={this.state.displayMsg} />
      }
    }
    return (
      <>
        {content}
        <ToastContainer autoClose={2000} />
      </>
    )
  }
}
