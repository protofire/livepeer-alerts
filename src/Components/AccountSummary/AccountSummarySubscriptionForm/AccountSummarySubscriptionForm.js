import * as displayTexts from '../AccountSummaryTexts'
import AccountSummarySubscriptionFormDisplay from './AccountSummaryFormDisplay/AccountSummarySubscriptionFormDisplay'
import React, { Component } from 'react'
import SpinnerExtended from '../../Common/UI/SpinnerExtended/SpinnerExtended'
import axios from 'axios'
import validator from 'validator'
import { toast, ToastContainer } from 'react-toastify'
import logdown from 'logdown'
import AccountSummaryModalEmail from './AccountSummaryFormDisplay/AccountSummaryModalEmail/AccountSummaryModalEmail'

const logger = logdown('Livepeer:AccountSummarySubscriptionForm')
logger.state.isEnabled = process.env.NODE_ENV !== 'production'

export class AccountSummarySubscriptionForm extends Component {
  state = {
    form: {
      email: {
        elementType: 'input',
        elementConfig: {
          type: 'email',
          placeholder: 'Your E-Mail'
        },
        validation: {
          required: true,
          emailCheck: true
        },
        value: '',
        touched: false,
        valid: false
      },
      formIsValid: false
    },
    address: null,
    frequency: 'daily',
    render: false,
    toastId: '1',
    displayMsg: displayTexts.LOADING_SUBSCRIPTION_DATA,
    userSubscribed: false
  }

  componentWillReceiveProps(nextProps, nextContext) {
    logger.log('Fire event componentWillReceiveProps')
    this.setState({
      ...this.state,
      address: nextProps.userData.address
    })
  }

  componentDidMount() {
    logger.log('Fire event componentDidMount')
    this.setState({
      address: this.props.userData.address,
      render: true
    })
  }

  onSubmitBtnHandler = async event => {
    event.preventDefault()
    logger.log('Submit btnHandler')
    let data = {
      address: this.state.address,
      frequency: this.state.frequency,
      email: this.state.form.email.value
    }
    this.setState(
      {
        render: false,
        displayMsg: displayTexts.GENERATING_SUBSCRIPTION
      },
      async () => {
        await this.generateSubscription(data, () => {
          //    this.props.history.push('/account')
        })
      }
    )
  }

  onCancelBtnHandler = event => {
    logger.log('Cancel btnHandler')
    this.props.history.push('/account')
  }

  onEmailModalClosed = event => {
    logger.log('Email modal closed')
    this.props.history.push('/account')
  }

  generateSubscription = async (data, callback) => {
    let response
    try {
      logger.log('Creating new subscriber with data: ', data)
      response = await axios.post('', data)
      this.setState(
        {
          userData: {
            ...this.props.userData,
            activated: response.data.activated,
            id: response.data._id,
            activatedCode: response.data.activated,
            createdAt: response.data.createdAt,
            isSubscribed: true
          },
          render: true,
          error: false,
          displayMsg: displayTexts.WELCOME_NEW_SUBSCRIBER
        },
        () => {
          setTimeout(callback, 1000)
        }
      )
    } catch (exception) {
      logger.log('Exception on postSubscription')
      let responseMsg = exception.response.data.message
      let displayMsg
      /** Email already exists **/
      if (
        (responseMsg && responseMsg === displayTexts.FAIL_EMAIL_ALREADY_EXISTS_RESPONSE) ||
        exception.response.status === 422
      ) {
        displayMsg = displayTexts.EMAIL_ALREADY_EXISTS
      } else {
        displayMsg = displayTexts.FAIL_NO_REASON
      }
      this.setState(
        {
          render: true,
          displayMsg: displayMsg,
          error: true
        },
        () => {
          this.sendToast()
        }
      )
    }
  }

  sendToast = (toastTime, callback) => {
    let time = 2000
    if (toastTime) {
      time = toastTime
    }
    let displayMsg = this.state.displayMsg

    if (!toast.isActive(this.state.toastId)) {
      if (this.state.error) {
        toast.error(displayMsg, {
          position: toast.POSITION.TOP_RIGHT,
          progressClassName: 'Toast-progress-bar',
          autoClose: time,
          toastId: this.state.toastId,
          onClose: callback
        })
      } else {
        toast.success(displayMsg, {
          position: toast.POSITION.TOP_RIGHT,
          progressClassName: 'Toast-progress-bar',
          autoClose: time,
          toastId: this.state.toastId,
          onClose: callback
        })
      }
    }
  }

  checkValidity = (value, rules) => {
    let isValid = true
    if (!rules) {
      return true
    }

    if (rules.required) {
      isValid = value.trim() !== '' && isValid
    }
    if (rules.minLength) {
      isValid = value.length >= rules.minLength && isValid
    }
    if (rules.maxLength) {
      isValid = value.length <= rules.maxLength && isValid
    }
    if (rules.emailCheck) {
      isValid = validator.isEmail(value) && isValid
    }
    return isValid
  }

  inputChangedHandler = (event, inputIdentifier) => {
    const updatedForm = {
      ...this.state.form
    }
    const updatedFormElement = {
      ...updatedForm[inputIdentifier],
      value: event.target.value
    }
    updatedFormElement.valid = this.checkValidity(
      updatedFormElement.value,
      updatedFormElement.validation
    )
    updatedFormElement.touched = true
    updatedForm[inputIdentifier] = updatedFormElement

    let formIsValid = true
    formIsValid = updatedForm[inputIdentifier].valid && formIsValid
    updatedForm.formIsValid = formIsValid
    this.setState({ form: updatedForm })
  }

  render() {
    let content = <SpinnerExtended displayMsg={this.state.displayMsg} />

    if (this.state.render) {
      if (this.state.userData && this.state.userData.isSubscribed) {
        console.log('acc summary')
        content = <AccountSummaryModalEmail onEmailModalClosed={this.onEmailModalClosed} />
      } else {
        content = (
          <AccountSummarySubscriptionFormDisplay
            form={this.state.form}
            onSubmitBtnHandler={this.onSubmitBtnHandler}
            onCancelBtnHandler={this.onCancelBtnHandler}
            inputChangedHandler={this.inputChangedHandler}
          />
        )
      }
    }

    return (
      <>
        {content}
        <ToastContainer autoClose={2000} />
      </>
    )
  }
}
