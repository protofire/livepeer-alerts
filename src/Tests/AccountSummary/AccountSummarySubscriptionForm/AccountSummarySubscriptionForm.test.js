import React from 'react'
import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import * as displayTexts from '../../../Components/AccountSummary/AccountSummaryTexts'
import { AccountSummarySubscriptionForm } from '../../../Components/AccountSummary/AccountSummarySubscriptionForm/AccountSummarySubscriptionForm'
import render from 'react-test-renderer'
import axiosInstance from '../../../util/axios'

configure({ adapter: new Adapter() })

const props = {
  userData: {
    address: '0x4d3F9184Fc32A43BAD2641b1536B52a076FBBDcE'
  }
}

const response = {
  data: {
    activated: true,
    _id: 1,
    createdAt: ''
  }
}
describe('AccountSummarySubscriptionForm Test', () => {
  it('Renders AccountSummarySubscriptionForm and match snapshot', () => {
    // Given

    let wrapper = render.create(<AccountSummarySubscriptionForm {...props} />)

    // When
    const tree = wrapper.toJSON()

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('Shows Welcome Message', () => {
    // Given
    const message = 'Subscribe:'

    //When
    let wrapper = mount(<AccountSummarySubscriptionForm {...props} />)

    //Then
    expect(wrapper.contains(message)).toEqual(true)
  })

  it('Renders Input email text', () => {
    // Given
    let wrapper = mount(<AccountSummarySubscriptionForm {...props} />)

    //When
    const input = wrapper.find('.InputElement')

    //Then
    expect(input.length).toEqual(1)
  })

  it('Renders Subscribe btn', () => {
    // Given
    let wrapper = mount(<AccountSummarySubscriptionForm {...props} />)

    //When
    const buttons = wrapper.find('button')

    //Then
    expect(buttons.length).toEqual(2)
  })

  it('Validity false is input no email', () => {
    //Given
    let wrapper = shallow(<AccountSummarySubscriptionForm {...props} />)
    const value = 'testnoemail'
    const rules = {
      required: true,
      emailCheck: true
    }

    //When
    const validity = wrapper.instance().checkValidity(value, rules)

    //Then
    expect(validity).toBe(false)
  })

  it('Validity true is input email', () => {
    // Given
    let wrapper = shallow(<AccountSummarySubscriptionForm {...props} />)
    const value = 'testemailok@test.com'
    const rules = {
      required: true,
      emailCheck: true
    }

    // When
    const validity = wrapper.instance().checkValidity(value, rules)

    // Then
    expect(validity).toBe(true)
  })

  it('Displays welcome subscriber if the user subscribed', async () => {
    let wrapper = mount(<AccountSummarySubscriptionForm {...props} />)
    const emailTest = 'test@test.com'
    axiosInstance.post = jest.fn().mockResolvedValue(response)
    let emailInput = wrapper.find('input').at(1)
    emailInput.simulate('change', { target: { value: emailTest } })
    await wrapper.instance().generateSubscription(null, null)
    let welcomeMsg = displayTexts.WELCOME_NEW_SUBSCRIBER
    expect(wrapper.state().displayMsg).toBe(welcomeMsg)
  })
})
