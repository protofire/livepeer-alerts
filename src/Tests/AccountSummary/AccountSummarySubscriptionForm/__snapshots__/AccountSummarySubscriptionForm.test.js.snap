// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSummarySubscriptionForm Test Renders AccountSummarySubscriptionForm and match snapshot 1`] = `
Array [
  <div
    className="MuiGrid-container-2 MuiGrid-align-items-xs-center-10 MuiGrid-justify-xs-center-19 GridContainer-grid-1 accountSummarySubscriptionFormDisplay-gridContainer-130"
  >
    <div
      className="MuiGrid-item-3 GridItem-grid-99 accountSummarySubscriptionFormDisplay-cardContainer-129"
    >
      <div
        className="Card-card-100 accountSummarySubscriptionFormDisplay-cardSignup-131"
      >
        <h2
          className="accountSummarySubscriptionFormDisplay-cardTitle-128"
        >
          Subscribe:
        </h2>
        <form>
          <div
            className="Input"
          >
            <label
              className="Label"
            />
            <input
              className="InputElement"
              onChange={[Function]}
              placeholder="Your E-Mail"
              type="email"
              value=""
            />
          </div>
          <button
            className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-text-164 MuiButton-flat-167 RegularButton-button-133 RegularButton-lg-156 RegularButton-warning-138 RegularButton-round-158 accountSummarySubscriptionFormDisplay-subscribeButton-132"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-163"
            >
              Cancel
            </span>
            <span
              className="MuiTouchRipple-root-191"
            />
          </button>
          <button
            className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-text-164 MuiButton-flat-167 RegularButton-button-133 RegularButton-lg-156 RegularButton-primary-135 RegularButton-round-158 RegularButton-disabled-155 accountSummarySubscriptionFormDisplay-subscribeButton-132"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-163"
            >
              Subscribe
            </span>
            <span
              className="MuiTouchRipple-root-191"
            />
          </button>
        </form>
      </div>
    </div>
  </div>,
  <div
    className="Toastify"
  />,
]
`;
